src/magic_folder/test/test_local_snapshot.py

```diff
@@ -0,0 +1,336 @@
+        return defer.succeed(None)
+
+
+class LocalSnapshotServiceTests(SyncTestCase):
+
+    def setUp(self):
+        super(LocalSnapshotServiceTests, self).setUp()
+        self.db = magicfolderdb.get_magicfolderdb(":memory:", create_version=(magicfolderdb.SCHEMA_v1, 1))
+        self.author = create_local_author("alice")
+
+        self.stash_dir = FilePath(self.mktemp())
+        self.stash_dir.makedirs()
+
+        self.magic_path = FilePath(self.mktemp())
+        self.magic_path.makedirs()
+
+        self._snapshot_creator = MemorySnapshotCreator()
+
+        self.snapshot_service = LocalSnapshotService(
+            magic_path=self.magic_path,
+            snapshot_creator=self._snapshot_creator,
+        )
+
+    def setup_example(self):
+        """
+        Hypothesis-invoked hook to create per-example state.
+        Reset the database before running each test.
+        """
+        self._snapshot_creator.processed = {}
+
+    def test_add_single_file(self):
+        """
+        Start the service, add a file and check if the operation succeeded.
+        """
+        foo = self.magic_path.child("foo")
+        content = u"foo"
+        with foo.open("w") as f:
+            f.write(content)
+
+        self.snapshot_service.startService()
+        d = self.snapshot_service.add_file(foo)
+
+        self.assertThat(
+            d,
+            succeeded(Always()),
+        )
+
+        self.assertThat(
+            self.snapshot_service.stopService(),
+            succeeded(Always())
+        )
+
+        self.assertThat(self._snapshot_creator.processed.get(foo), Not(Equals(None)))
+        self.assertThat(self._snapshot_creator.processed.get(foo), Equals(0))
+
+    @given(lists(path_segments().map(lambda p: p.encode("utf-8")), unique=True),
+           lists(binary(), unique=True))
+    def test_add_multiple_files(self, filenames, contents):
+        """
+        Add a bunch of files one by one and check whether the operation is
+        successful.
+        """
+        files = []
+        for (filename, content) in zip(filenames, contents):
+            file = self.magic_path.child(filename)
+            with file.open("wb") as f:
+                f.write(content)
+            files.append(file)
+
+        self.snapshot_service.startService()
+
+        list_d = []
+        for file in files:
+            result_d = self.snapshot_service.add_file(file)
+            list_d.append(result_d)
+
+        d = defer.gatherResults(list_d)
+
+        self.assertThat(
+            d,
+            succeeded(Always()),
+        )
+
+        self.assertThat(
+            self.snapshot_service.stopService(),
+            succeeded(Always())
+        )
+
+        self.assertThat(
+            len(self._snapshot_creator.processed.keys()),
+            Equals(len(files))
+        )
+
+        self.assertThat(
+            sorted(self._snapshot_creator.processed.keys()),
+            Equals(sorted(files))
+        )
```

Comment:
If the processed dict's keys equal the sorted files list then their lengths must be equal, also, so the length assertion seems redundant.